### PR TITLE
fault: use FractionalPercent for percent

### LINKED ISSUE #3904: Fault Delay: support float type percent
*Title*: *Fault Delay: support float type percent*

*Description*:
The `percent` field in the config of Fault Delay is limited to [uint32](https://www.envoyproxy.io/docs/envoy/v1.7.0/api-v2/config/filter/fault/v2/fault.proto.html?highlight=fault%20delay).
https://github.com/envoyproxy/envoy/blob/1fdb6386c4bb42748530d7a9bf58ded644d77749/api/envoy/config/filter/fault/v2/fault.proto#L27
However, for upstream clusters with heavy traffic, even 1% can be an exceedingly large number of requests. Would be useful to be able to specify percent as a float, e.g., 0.001%

### PR DESCRIPTION
**Pull request by venilnoronha**
*Description*:
The existing `FaultDelay` config used `uint32` for the `percent` field which limited the user to configure percentages only in terms of whole numbers between 0 and 100. `FractionalPercent` allows finer control over the percent values by allowing fractions to be specified in the configuration, for example, 0.0001%.

Signed-off-by: Venil Noronha <veniln@vmware.com>

*Risk Level*: Med
*Testing*: Added 2 new tests and updated existing tests
*Docs Changes*: N/A
*Release Notes*: Added deprecation note and introduced `FractionalPercent` percentages.
Fixes #3904

/cc @rshriram

### CODE REVIEW

File: api/envoy/config/filter/fault/v2/fault.proto [hunk 1]
```diff
@@ -22,9 +24,8 @@ message FaultDelay {
   // supported.
   FaultDelayType type = 1 [(validate.rules).enum.defined_only = true];
 
-  // An integer between 0-100 indicating the percentage of operations/connection requests
-  // on which the delay will be injected.
-  uint32 percent = 2 [(validate.rules).uint32.lte = 100];
+  // The percentage of operations/connection requests on which the delay will be injected.
+  envoy.type.FractionalPercent percent = 2;
```
Comment (mattklein123) [inline api/envoy/config/filter/fault/v2/fault.proto]: We can't just change this. You will need to deprecate the old percent and add a new one.

File: api/envoy/type/percent.proto [hunk 2]
```diff
@@ -31,15 +31,25 @@ message FractionalPercent {
...
 
+    // 1,000.
+    //
+    // **Example**: 1/1000 = 0.1%.
+    THOUSAND = 1;
+
     // 10,000.
     //
     // **Example**: 1/10000 = 0.01%.
-    TEN_THOUSAND = 1;
```
Comment (mattklein123) [inline api/envoy/type/percent.proto]: These can't be changed. You will need to add new ones at the end. Are they really necessary though? Seems unnecessary to me with TEN_THOUSAND and MILLION. I would revert these changes personally.

File: docs/root/api-v1/http_filters/fault_filter.rst [hunk 3]
```diff
@@ -77,7 +77,10 @@ Delay
 
   {
     "type" : "...",
-    "fixed_delay_percent" : "...",
+    "fixed_delay_percent" : {
```
Comment (mattklein123) [inline docs/root/api-v1/http_filters/fault_filter.rst]: We don't accept v1 changes anymore. Please revert all changes to v1 code, schema, docs, etc.

**Comment by venilnoronha** [comment]
@mattklein123 thanks for the quick review! I'll address the issues you mentioned.

File: api/envoy/type/percent.proto (hunk 2 above)
Comment (venilnoronha) [reply api/envoy/type/percent.proto]: Addressed in f105661.

**Comment by mattklein123** [comment]
@rshriram do you mind doing first pass review on this?

File: api/envoy/config/filter/fault/v2/fault.proto (hunk 1 above)
Comment (rshriram) [reply api/envoy/config/filter/fault/v2/fault.proto]: call the new one as percentage. And i think there is also some protoc validation with it

File: docs/root/api-v1/http_filters/fault_filter.rst [hunk 4]
```diff
@@ -86,9 +89,8 @@ type:
   injected. Currently only *fixed* delay type (step function) is supported.
 
 fixed_delay_percent:
-  *(required, integer)* The percentage of requests that will
-  be delayed for the duration specified by *fixed_duration_ms*. Valid
-  values range from 0 to 100.
+  *(required, envoy.type.FractionalPercent)* The percentage of requests that will be delayed for
+  the duration specified by *fixed_duration_ms*.
```
Comment (rshriram) [inline docs/root/api-v1/http_filters/fault_filter.rst]: Skip all changes to this file

File: docs/root/api-v1/network_filters/mongo_proxy_filter.rst [hunk 5]
```diff
@@ -39,14 +39,17 @@ incoming data up until the timer event fires will be a part of the delay.
...
+      },
       "duration_ms": "..."
     }
   }
 
 percent
-  *(required, integer)* Probability of an eligible MongoDB operation to be affected by the
-  injected fault when there is no active fault. Valid values are integers in a range of [0, 100].
+  *(required, envoy.type.FractionalPercent)* Probability of an eligible MongoDB operation to be
+  affected by the injected fault when there is no active fault.
```
Comment (rshriram) [inline docs/root/api-v1/network_filters/mongo_proxy_filter.rst]: and this as well

File: include/envoy/runtime/runtime.h [hunk 6]
```diff
@@ -112,12 +125,12 @@ class Snapshot {
...
-   * @param control max number of buckets for sampling. Sampled value will be in a range of
-   *        [0, num_buckets).
+   * @param num_buckets control max number of buckets for sampling. Sampled value will be in a range
+   *        of [0, num_buckets).
    * @return true if the feature is enabled.
    */
-  virtual bool featureEnabled(const std::string& key, uint64_t default_value, uint64_t random_value,
-                              uint64_t num_buckets) const PURE;
+  virtual bool sampleFeatureEnabled(const std::string& key, uint64_t default_value,
+                                    uint64_t random_value, uint64_t num_buckets) const PURE;
```
Comment (rshriram) [inline include/envoy/runtime/runtime.h]: @mattklein123 how are you guys doing fractional percent with runtimes? IOW, do you need runtime support for fractional percent? If not, all changes to this file can be eliminated

File: source/common/access_log/access_log_impl.cc [hunk 7]
```diff
@@ -118,7 +118,7 @@ bool RuntimeFilter::evaluate(const RequestInfo::RequestInfo&,
     random_value = random_.random();
   }
 
-  return runtime_.snapshot().featureEnabled(
+  return runtime_.snapshot().sampleFeatureEnabled(
```
Comment (rshriram) [inline source/common/access_log/access_log_impl.cc]: For the moment, undo all changes to featureEnabled(). Lets not support runtimes for the fractional percent. You can simply do a if percent != 0 (old path) else (new path)

File: source/common/json/config_schemas.cc [hunk 8]
```diff
@@ -379,9 +379,20 @@ const std::string Json::Schema::MONGO_PROXY_NETWORK_FILTER_SCHEMA(R"EOF(
             "type" : "object",
             "properties" : {
               "percent" : {
-                "type" : "integer",
-                "minimum" : 0,
-                "maximum" : 100
```
Comment (rshriram) [inline source/common/json/config_schemas.cc]: skip file

File: include/envoy/runtime/runtime.h (hunk 6 above)
Comment (venilnoronha) [reply include/envoy/runtime/runtime.h]: I think `access_log_impl` uses `runtime` for sampling the feature status as seen [here](https://github.com/envoyproxy/envoy/blob/2a68458119ffcd97498fcbe13fc14c8b8d043cb0/source/common/access_log/access_log_impl.cc#L121-L123). Should we just stick to a similar approach for `fault`s?

Comment (rshriram) [reply include/envoy/runtime/runtime.h]: Looking again, why is this being renamed ?

Comment (venilnoronha) [reply include/envoy/runtime/runtime.h]: Because of a signature conflict between the routines highlighted [here](
https://github.com/venilnoronha/envoy/blob/f1056616cfb0b87e1b0be75102f89e452955a664/include/envoy/runtime/runtime.h).

File: source/common/runtime/runtime_impl.cc [hunk 9]
```diff
@@ -146,8 +146,13 @@ std::string RandomGeneratorImpl::uuid() {
...
 }
 
-bool SnapshotImpl::featureEnabled(const std::string& key, uint64_t default_value,
-                                  uint64_t random_value, uint64_t num_buckets) const {
+bool SnapshotImpl::sampleFeatureEnabled(const std::string& key, uint64_t default_value,
+                                        uint64_t num_buckets) const {
+  return sampleFeatureEnabled(key, default_value, generator_.random(), num_buckets);
+}
+
+bool SnapshotImpl::sampleFeatureEnabled(const std::string& key, uint64_t default_value,
```
[the remark this answers is not in the document]
Comment (rshriram) [reply source/common/runtime/runtime_impl.cc]: Ignore

File: source/extensions/filters/http/fault/fault_filter.h [hunk 10]
```diff
@@ -47,17 +47,19 @@ class FaultSettings : public Router::RouteSpecificFilterConfig {
...
 private:
-  uint64_t abort_percent_{};       // 0-100
-  uint64_t http_status_{};         // HTTP or gRPC return codes
-  uint64_t fixed_delay_percent_{}; // 0-100
-  uint64_t fixed_duration_ms_{};   // in milliseconds
-  std::string upstream_cluster_;   // restrict faults to specific upstream cluster
+  uint64_t abort_percent_{};                                // 0-100
+  uint64_t http_status_{};                                  // HTTP or gRPC return codes
+  uint64_t fixed_delay_percent_{};                          // 0-100
+  envoy::type::FractionalPercent fixed_delay_percentage_{}; // 0-100
```
[the remark this answers is not in the document]
Comment (venilnoronha) [reply source/extensions/filters/http/fault/fault_filter.h]: I'd prefer to implement the change to abort in a separate PR.

File: source/extensions/filters/http/fault/fault_filter.cc [hunk 11]
```diff
@@ -41,6 +40,7 @@ FaultSettings::FaultSettings(const envoy::config::filter::http::fault::v2::HTTPF
 
   if (fault.has_delay()) {
     fixed_delay_percent_ = fault.delay().percent();
+    fixed_delay_percentage_ = fault.delay().percentage();
```
Comment (rshriram) [inline source/extensions/filters/http/fault/fault_filter.cc]: do you have validtion set to ensure that this field is not null ?
Also, fold fixed_delay_percent_ into fixed_delay_percentage.. i.e. convert the old percent (if set) to new form..
The problem though is that you can't differentiate between a 0 percent and unset percent in hte old one as its uint32.. Sigh..

Comment (venilnoronha) [reply source/extensions/filters/http/fault/fault_filter.cc]: So, we keep this as is with the validation brought in?

Comment (venilnoronha) [reply source/extensions/filters/http/fault/fault_filter.cc]: I've addressed this in 6652866.

File: source/extensions/filters/http/fault/fault_filter.h [hunk 12]
```diff
@@ -46,18 +46,18 @@ class FaultSettings : public Router::RouteSpecificFilterConfig {
...
   const std::string& upstreamCluster() const { return upstream_cluster_; }
   const std::unordered_set<std::string>& downstreamNodes() const { return downstream_nodes_; }
 
 private:
-  uint64_t abort_percent_{};       // 0-100
-  uint64_t http_status_{};         // HTTP or gRPC return codes
-  uint64_t fixed_delay_percent_{}; // 0-100
-  uint64_t fixed_duration_ms_{};   // in milliseconds
-  std::string upstream_cluster_;   // restrict faults to specific upstream cluster
+  uint64_t abort_percent_{};                                // 0-100
```
[the remark this answers is not in the document]
Comment (venilnoronha) [reply source/extensions/filters/http/fault/fault_filter.h]: Since this PR is big in itself, should we do it in another PR?

**Comment by venilnoronha** [comment]
@rshriram I've addressed the last set of comments in f14d3bc. Let me know if you still want the abort changes in this PR, or if you'd prefer it in another PR.

File: source/extensions/filters/http/fault/fault_filter.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (rshriram) [reply source/extensions/filters/http/fault/fault_filter.h]: The abort stuff is small. Confined to http only. If done right, it should be touching same files and not much.

**Comment by rshriram** [comment]
Same pr please. It’s a mass search and replace. Makes it easy to follow patterns

File: source/extensions/filters/http/fault/fault_filter.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (venilnoronha) [reply source/extensions/filters/http/fault/fault_filter.h]: Yes, I'm on it right now. Figured out it wasn't as big of a change as I thought it would be.

**Comment by venilnoronha** [comment]
Implemented the changes to abort in 3d56b3f8.

**Comment by rshriram** [comment]
ping @mattklein123 .. this is ready for a final pass.

File: docs/root/intro/version_history.rst [hunk 13]
```diff
@@ -11,6 +11,7 @@ Version history
   `google.api.HttpBody <https://github.com/googleapis/googleapis/blob/master/google/api/httpbody.proto>`_.
 * config: v1 disabled by default. v1 support remains available until October via flipping --v2-config-only=false.
 * config: v1 disabled by default. v1 support remains available until October via setting :option:`--allow-deprecated-v1-api`.
+* fault: added support for :ref:`fractional percentages <envoy_api_field_config.filter.fault.v2.FaultDelay.percentage>`.
```
Comment (mattklein123) [inline docs/root/intro/version_history.rst]: Please note the deprecations in deprecated.md

File: include/envoy/runtime/runtime.h [hunk 14]
```diff
@@ -102,6 +102,19 @@ class Snapshot {
...
+   * Test if a feature is enabled using the built in random generator and total number of buckets
+   * for sampling.
+   * @param key supplies the feature key to lookup.
+   * @param default_value supplies the default value that will be used if either the feature key
+   *        does not exist or it is not an integer.
+   * @param num_buckets control max number of buckets for sampling. Sampled value will be in a range
+   * of [0, num_buckets).
+   * @return true if the feature is enabled.
+   */
+  virtual bool sampleFeatureEnabled(const std::string& key, uint64_t default_value,
```
Comment (mattklein123) [inline include/envoy/runtime/runtime.h]: nit: I see why you had to change the name here, but IMO "sample" is a bit strange in the sense that the other featureEnabled() functions also do sampling. A few options here:
1) Get rid of the versions of featureEnabled() that don't require specifying number of buckets. Larger change.
2) Some other name. `featureEnabledEx` ? (Bad, but not sure what would be better).

File: source/extensions/filters/http/fault/fault_filter.cc [hunk 15]
```diff
@@ -34,12 +34,22 @@ const std::string FaultFilter::ABORT_HTTP_STATUS_KEY = "fault.http.abort.http_st
 FaultSettings::FaultSettings(const envoy::config::filter::http::fault::v2::HTTPFault& fault) {
 
   if (fault.has_abort()) {
-    abort_percent_ = fault.abort().percent();
+    if (fault.abort().has_percentage()) {
```
Comment (mattklein123) [inline source/extensions/filters/http/fault/fault_filter.cc]: extract to helper function for duplicated code here and below.

File: source/extensions/filters/http/fault/fault_filter.h [hunk 16]
```diff
@@ -45,19 +45,19 @@ class FaultSettings : public Router::RouteSpecificFilterConfig {
...
   const std::string& upstreamCluster() const { return upstream_cluster_; }
   const std::unordered_set<std::string>& downstreamNodes() const { return downstream_nodes_; }
 
 private:
-  uint64_t abort_percent_{};       // 0-100
-  uint64_t http_status_{};         // HTTP or gRPC return codes
-  uint64_t fixed_delay_percent_{}; // 0-100
-  uint64_t fixed_duration_ms_{};   // in milliseconds
-  std::string upstream_cluster_;   // restrict faults to specific upstream cluster
+  envoy::type::FractionalPercent abort_percentage_{};       // 0.0-100.0
```
Comment (mattklein123) [inline source/extensions/filters/http/fault/fault_filter.h]: `{}` not needed. Also, comment is no longer relevant I think. Same below.

File: source/extensions/filters/network/mongo_proxy/proxy.h [hunk 17]
```diff
@@ -102,13 +102,19 @@ typedef std::shared_ptr<AccessLog> AccessLogSharedPtr;
 class FaultConfig {
 public:
   FaultConfig(const envoy::config::filter::fault::v2::FaultDelay& fault_config)
-      : delay_percent_(fault_config.percent()),
-        duration_ms_(PROTOBUF_GET_MS_REQUIRED(fault_config, fixed_delay)) {}
-  uint32_t delayPercent() const { return delay_percent_; }
+      : duration_ms_(PROTOBUF_GET_MS_REQUIRED(fault_config, fixed_delay)) {
+    if (fault_config.has_percentage()) {
```
Comment (mattklein123) [inline source/extensions/filters/network/mongo_proxy/proxy.h]: same utility function extracted above

File: include/envoy/runtime/runtime.h (hunk 14 above)
Comment (venilnoronha) [reply include/envoy/runtime/runtime.h]: @mattklein123 thanks for reviewing! I've addressed your other comments in f197fe2, and I'm working on this now.

Comment (venilnoronha) [reply include/envoy/runtime/runtime.h]: Implemented option 1 in f98bcc0.

File: source/common/http/conn_manager_utility.cc [hunk 18]
```diff
@@ -212,7 +212,7 @@ void ConnectionManagerUtility::mutateTracingRequestHeader(Http::HeaderMap& reque
   }
 
   if (!runtime.snapshot().featureEnabled("tracing.global_enabled",
-                                         config.tracingConfig()->overall_sampling_, result)) {
+                                         config.tracingConfig()->overall_sampling_, result, 100)) {
```
Comment (rshriram) [inline source/common/http/conn_manager_utility.cc]: why this change?

Comment (venilnoronha) [reply source/common/http/conn_manager_utility.cc]: See [this comment](
https://github.com/envoyproxy/envoy/pull/3978#discussion_r208368275)

**Comment by rshriram** [comment]
@venilnoronha I think you lost some of your changes. You haven't implemented either option that matt suggested but the call sites seem to have changed

File: include/envoy/runtime/runtime.h [hunk 19]
```diff
@@ -90,17 +90,17 @@ class Snapshot {
...
    *        does not exist or it is not an integer.
-   * @param random_value supplies the stable random value to use for determining whether the feature
-   *        is enabled.
+   * @param num_buckets control max number of buckets for sampling. Sampled value will be in a range
+   *        of [0, num_buckets).
    * @return true if the feature is enabled.
    */
   virtual bool featureEnabled(const std::string& key, uint64_t default_value,
-                              uint64_t random_value) const PURE;
+                              uint64_t num_buckets) const PURE;
```
Comment (rshriram) [inline include/envoy/runtime/runtime.h]: did you intend to delete this function and simply use the one below everywhere?

**Comment by rshriram** [comment]
So I suggest this.. Replace all call sites of featureEnabled() with the longer version that always passes the number of buckets - But do this in a separate PR. This should essentially be a mass search-and-replace. Should have no issues with any reviewer. Then this PR will automatically reduce to the changes related to fault injection.

Alternatively, simply rename the sampleFeatureEnabled to something else like featureEnabledEx and retain changes before you followed matt's option

File: source/common/runtime/runtime_impl.cc [hunk 20]
```diff
@@ -163,11 +168,6 @@ bool SnapshotImpl::featureEnabled(const std::string& key, uint64_t default_value
   }
 }
 
-bool SnapshotImpl::featureEnabled(const std::string& key, uint64_t default_value,
-                                  uint64_t random_value) const {
-  return featureEnabled(key, default_value, random_value, 100);
-}
-
```
[the remark this answers is not in the document]
Comment (venilnoronha) [reply source/common/runtime/runtime_impl.cc]: `random_value` and `num_buckets` aren't being used interchangeably. Either we [pass both `random_value` and `num_buckets` to `featureEnabled`](https://github.com/envoyproxy/envoy/pull/3978/files#diff-2f4944a580a2633705f842bf4ff9c7f9R120), or we [only pass `num_buckets`](https://github.com/envoyproxy/envoy/pull/3978/files#diff-2f4944a580a2633705f842bf4ff9c7f9R102). There's no way to pass `random_value` without `num_buckets`.

**Comment by venilnoronha** [comment]
@rshriram I've reverted the changes and have renamed `sampleFeatureEnabled` to `featureEnabledEx` in 9b1046f.

@mattklein123 as @rshriram suggested, I'll take up [option 1](https://github.com/envoyproxy/envoy/pull/3978#discussion_r208368275) in a separate PR.

Thanks,
Venil

**Comment by mattklein123** [comment]
> @mattklein123 as @rshriram suggested, I'll take up option 1 in a separate PR.

I would rather do one set of changes. If you want to do option 1, can you just do that PR first, we can review/merge, and then we can merge master into this?

**Comment by venilnoronha** [comment]
On a second thought, I'd prefer keeping `featureEnabledEx`.

File: include/envoy/runtime/runtime.h [hunk 21]
```diff
@@ -102,6 +102,19 @@ class Snapshot {
...
+   * for sampling.
+   * @param key supplies the feature key to lookup.
+   * @param default_value supplies the default value that will be used if either the feature key
+   *        does not exist or it is not an integer.
+   * @param num_buckets control max number of buckets for sampling. Sampled value will be in a range
+   * of [0, num_buckets).
+   * @return true if the feature is enabled.
+   */
+  virtual bool featureEnabledEx(const std::string& key, uint64_t default_value,
+                                uint64_t num_buckets) const PURE;
```
Comment (rshriram) [inline include/envoy/runtime/runtime.h]: Is there any callsite that is calling the 3 arg form of featureEnabledEx supplying buckets instead of random value? If not, lets get rid of this function.

Comment (mattklein123) [reply include/envoy/runtime/runtime.h]: +1 if possible.

Comment (venilnoronha) [reply include/envoy/runtime/runtime.h]: Addressed via aef3da0 and c31bd82.

File: include/envoy/runtime/runtime.h [hunk 22]
```diff
@@ -165,6 +165,11 @@ class Loader {
    * @param values the values to merge
    */
   virtual void mergeValues(const std::unordered_map<std::string, std::string>& values) PURE;
+
+  /**
+   * @return RandomGenerator& the random generator.
+   */
+  virtual RandomGenerator& random() PURE;
```
Comment (mattklein123) [inline include/envoy/runtime/runtime.h]: Do we really need to add this to this interface? I guess it's OK, but can't you just pass the random generator to where you need it? It should be in the filter config factory? This is what we do elsewhere.

Comment (venilnoronha) [reply include/envoy/runtime/runtime.h]: Thought that I'd have to pass the `RandomGenerator` instance a long way, but I was wrong. I've fixed this in 54c6ac0.

**Comment by venilnoronha** [comment]
@mattklein123 @rshriram thanks for reviewing/merging.
